keps/sig-api-machinery/20200309-consistent-resource-versions-semantics.md

```diff
@@ -0,0 +1,216 @@
+	//   - if non-zero and Limit is set, ResourceVersionMatchExact rules applies implicitly.
+	// +optional
+	ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,4,opt,name=resourceVersion"`
+
+	// ResourceVersionMatch determines how ResourceVersion is applied. Not supported for watch calls.
+	// ResourceVersionMatch SHOULD be set for list calls where ResourceVersion is set. If ResourceVersion is unset,
+	// ResourceVersionMatch is ignored.
+	// For backward compatibility, clients must tolerate the server ignoring ResourceVersionMatch:
+	// - When using ResourceVersionMatchNotOlderThan and Limit is set, clients must handle HTTP 410 “Gone” responses.
+	//   For example, the client might retry with a newer ResourceVersion or fall back to a ResourceVersion="" request.
+	// - When using ResourceVersionMatchExact and Limit is unset, clients must verify that the ResourceVersion in the
+	//   ListMeta of the response matches the requested ResourceVersion, and handle the case where it does not. For
+	//   example, the client might fall back to the a request with Limit set.
+	// +optional
+	ResourceVersionMatch ResourceVersionMatch `json:"resourceVersionMatch,omitempty" protobuf:"bytes,10,opt,name=resourceVersionMatch"`
+    ...
+}
+```
+
+### Backward Compatibility
+
+Versions of the kube-apiserver that pre-date the introduction of `ResourceVersionMatch`
+parameter will ignore it. Client will need to tolerate this as documented on ResourceVersionWatch:
+
+- When using ResourceVersionMatchNotOlderThan and Limit is set, clients must handle HTTP 410 “Gone” responses.
+  For example, the client might retry with a newer ResourceVersion or fall back to a ResourceVersion="" request.
+- When using ResourceVersionMatchExact and Limit is unset, clients must verify that the ResourceVersion in the
+  ListMeta of the response matches the requested ResourceVersion, and handle the case where it does not. For
+  example, the client might fall back to the a request with Limit set.
+
+When 'ResourceVersionMatch' is not provided, the behavior is the same as before it was introduced.
+
+### Get support?
+
+`ResourceVersionMatch` can also be added to the get operation for consistency. `Get` currently
+`NotOlderThan` semantics by default and this would add support for `Exact`.
+
+In order to be backward compatible, get responses must include the `ResourceVersion` that the request was served at.
+For list responses this is provided in `ListMeta`, but get responses do not have a wrapper object like `ListMeta`.
+The `ObjectMeta.ResourceVersion` cannot be used because it represents the resource version that the object was
+created or last modified at, not the resource version is was served from.
+
+Options:
+- Don't support `ResourceVersionMatch` for get.
+- Add a header that provides the `ResourceVersion` the get was served at back to clients, e.g. `ServedAtResourceVersion: 43049`
+
+I am currently considering not including `ResourceVersionMatch` since:
```

Comment:
I'd change this section to say that get is not going to be this flexible. If people need particular semantics they can do a list and set `fieldSelector=metadata.name=foo`.

Reply:
+1 - that path is already optimized in apiserver (MatchesSingle() [or sth like that] thing), - so yeah, I agree that it's too high cost to understand it for not much gain.

Reply:
Great. I've rewritten this section to summarize why we're leaving get as-is.